README.md


<div id="badges"><!-- pkgdown markup -->
<a href="https://bioconductor.org/packages/illuminaio/"><img border="0" src="https://bioconductor.org/shields/years-in-bioc/illuminaio.svg" alt="Bioconductor since badge"/></a> <a href="https://bioconductor.org/checkResults/release/bioc-LATEST/illuminaio/"><img border="0" src="https://bioconductor.org/shields/build/release/bioc/illuminaio.svg" alt="Bioconductor build status"/></a> <a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/illuminaio/"><img border="0" src="https://bioconductor.org/shields/build/devel/bioc/illuminaio.svg" alt="Bioconductor build status"/></a> <a href="https://github.com/HenrikBengtsson/illuminaio/actions?query=workflow%3AR-CMD-check"><img border="0" src="https://github.com/HenrikBengtsson/illuminaio/actions/workflows/R-CMD-check.yaml/badge.svg?branch=develop" alt="R CMD check status"/></a>     <a href="https://app.codecov.io/gh/HenrikBengtsson/illuminaio"><img border="0" src="https://codecov.io/gh/HenrikBengtsson/illuminaio/branch/develop/graph/badge.svg" alt="Coverage Status"/></a> 
</div>

# illuminaio: Parsing Illumina Microarray Output Files 


## Installation
R package illuminaio is available on [Bioconductor](https://www.bioconductor.org/packages/devel/bioc/html/illuminaio.html) and can be installed in R as:

```r
install.packages("BiocManager")
BiocManager::install("illuminaio")
```


### Pre-release version

To install the pre-release version that is available in Git branch `develop` on GitHub, use:
```r
remotes::install_github("HenrikBengtsson/illuminaio", ref="develop")
```
This will install the package from source.  Because of this and because this package also compiles native code, Windows users need to have [Rtools](https://cran.r-project.org/bin/windows/Rtools/) installed and macOS users need to have [Xcode](https://developer.apple.com/xcode/) installed.


<!-- pkgdown-drop-below -->


## Contributing

To contribute to this package, please see [CONTRIBUTING.md](CONTRIBUTING.md).



NEWS
Package: illuminaio
===================

Version: 0.39.0-9000 [2022-04-26]

 * ...
 

Version: 0.39.0 [2022-04-26]

 * The version number was bumped for the Bioconductor devel version,
   which is now BioC 3.16 for R-devel.


Version: 0.38.0 [2022-04-26]

 * The version number was bumped for the Bioconductor release version,
   which is now BioC 3.15 for R (>= 4.2.0).


Version: 0.37.0 [2021-10-27]

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC 3.15 for R-devel.


Version: 0.36.0 [2021-10-27]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.14 for R (>= 4.1.1).


Version: 3.24.0 [2021-05-19]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.13 for R (>= 4.0.3).


Version: 3.22.0 [2020-10-27]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.12 for R (>= 4.0.0).


Version: 0.31.1 [2020-07-15]

UPDATES:

 * Error messages produced by readIDAT() when failing decrypt no longer appends
   an extra newline at the end.

 * readIDAT() uses explicit stringsAsFactors=FALSE internally.

 * readIDAT() no longer keeps two file connections open at the same.
 
DOCUMENTATION:

 * The BibTeX URL reported by citation(package="illuminaio") was broken.

BUG FIXES:

 * readBGX() would leave an open connection if there was an file reading error.
 

Version: 0.31.0 [2020-04-27]

NOTES:

 * The version number was bumped for the Bioconductor develop version, which is
   now BioC 3.12 for R (>= 4.0.0).


Version: 0.30.0 [2020-04-27]

NOTES:

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.11 for R (>= 4.0.0).


Version: 0.28.0 [2019-10-29]

NOTES:

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.10 for R (>= 3.6.1).


Version: 0.27.1 [2019-08-19]

SOFTWARE QUALITY:

 * Update decryption code to address unused result warning.
 

Version: 0.27.0 [2019-05-02]

NOTES:

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC 3.10 for R (>= 3.6.0).


Version: 0.26.0 [2019-05-02]

NOTES:

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.9 for R (>= 3.6.0).


Version: 0.25.0 [2018-10-30]

NOTES:

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC 3.9 for R (>= 3.6.0).


Version: 0.24.0 [2018-10-30]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.8 for R (>= 3.5.1).


Version: 0.23.2 [2018-07-18]

NEW FEATURES:

 * The readIDAT() function gains a 'what' argument (only for unencrypted IDAT
   files).  This allows the fast return of the number of 'nSNPsRead' (number
   of probes in file) and 'IlluminaID' (probenames). This is to allow fast
   handling of IDAT files in the minfi package.
   

Version: 0.23.1 [2018-07-18]

SOFTWARE QUALITY:

 * ROBUSTNESS: Now registering native routines.


Version: 0.23.0 [2018-05-01]

NOTES:

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC v3.8 for R (>= 3.5.0).


Version: 0.22.0 [2017-05-01]

NOTES:

 * The version number was bumped for the Bioconductor release version, which is
   now BioC v3.7 for R (>= 3.5.0).


Version: 0.21.0 [2017-10-31]

NOTES:

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC v3.7 for R (>= 3.5.0).


Version: 0.20.0 [2017-10-31]

NOTES:

 * The version number was bumped for the Bioconductor release version, which is
   now BioC v3.6 for R (>= 3.4.0).


Version: 0.19.0 [2017-04-25]

NOTES:

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC v3.6 for R (>= 3.4.0).


Version: 0.18.0 [2017-04-25]

NOTES:

 * The version number was bumped for the Bioconductor release version, which is
   now BioC v3.5 for R (>= 3.4.0).


Version: 0.17.0 [2016-10-18]

NOTES:

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC v3.5 for R (>= 3.4.0).


Version: 0.16.0 [2016-10-18]

NOTES:

 * The version number was bumped for the Bioconductor release version, which is
   now BioC v3.4 for R (>= 3.3.1).


Version 0.15.1 [2016-08-27]

 * Now the package DLL is unloaded when the package is unloaded.


Version: 0.15.0 [2015-05-03]

NOTES:

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC v3.4 for R (>= 3.3.0).


Version: 0.14.0 [2015-05-03]

NOTES:

 * The version number was bumped for the Bioconductor release version, which is
   now BioC v3.3 for R (>= 3.3.0).


Version 0.13.1 [2016-01-12]

DEPRECATED AND DEFUNCT:

 * readBPM() is deprecated because it was only a stub that never really worked.
   (Issues #5 and #6)


Version: 0.13.0 [2015-10-23]

NOTES:

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC v3.3 for R (>= 3.3.0).


Version: 0.12.0 [2015-10-13]

NOTES:

 * The version number was bumped for the Bioconductor release version, which is
   now BioC v3.2 for R (>= 3.2.2).


Version: 0.11.2 [2015-09-11]

BUG FIXES:

 * readIDAT() can now read non-encrypted IDAT files with strings longer than
   127 characters.

 * readIDAT() incorrectly assumed that there were exactly two blocks in RunInfo
   fields of non-encrypted (v3) IDAT files.  Thanks to Gordon Bean (GitHub
   @brazilbean) for reporting on and contributing with code for the above two
   bugs. (Issue #2)


Version: 0.11.1 [2015-07-29]

NOTES:

 * Updated the BiocViews field of DESCRIPTION.


Version: 0.11.0 [2015-04-16]

NOTES:

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC v3.2 for R (>= 3.3.0).


Version: 0.10.0 [2015-04-16]


src/des.c
/* des.c --- DES and Triple-DES encryption/decryption Algorithm
 * Copyright (C) 1998, 1999, 2001, 2002, 2003, 2004, 2005, 2006, 2007
 *    Free Software Foundation, Inc.
 *
 * This file is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2, or (at your
 * option) any later version.
 *
 * This file is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this file; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 *
 */

/* Adapted for gnulib by Simon Josefsson, based on Libgcrypt. */

/* Adapted for illuminaio by Mike Smith. */

/*
 *  U S A G E
 * ===========
 *
 * For DES or Triple-DES encryption/decryption you must initialize a proper
 * encryption context with a key.
 *
 * A DES key is 64bit wide but only 56bits of the key are used. The remaining
 * bits are parity bits and they will _not_ checked in this implementation, but
 * simply ignored.
 *
 * For Triple-DES you could use either two 64bit keys or three 64bit keys.
 * The parity bits will _not_ checked, too.
 *
 * After initializing a context with a key you could use this context to
 * encrypt or decrypt data in 64bit blocks in Electronic Codebook Mode.
 *
 * DES Example
 * -----------
 *     unsigned char key[8];
 *     unsigned char plaintext[8];
 *     unsigned char ciphertext[8];
 *     unsigned char recoverd[8];
 *     gl_des_ctx context;
 *
 *     // Fill 'key' and 'plaintext' with some data
 *     ....
 *
 *     // Set up the DES encryption context
 *     gl_des_setkey(&context, key);
 *
 *     // Encrypt the plaintext
 *     des_ecb_encrypt(&context, plaintext, ciphertext);
 *
 *     // To recover the orginal plaintext from ciphertext use:
 *     des_ecb_decrypt(&context, ciphertext, recoverd);
 *
 *
 */


//#include <config.h>

#include "des.h"

#include <stdio.h>
#include <string.h>             /* memcpy, memcmp */

/*
 * The s-box values are permuted according to the 'primitive function P'
 * and are rotated one bit to the left.
 */
static const uint32_t sbox1[64] = {
  0x01010400, 0x00000000, 0x00010000, 0x01010404, 0x01010004, 0x00010404,
  0x00000004, 0x00010000, 0x00000400, 0x01010400, 0x01010404, 0x00000400,
  0x01000404, 0x01010004, 0x01000000, 0x00000004, 0x00000404, 0x01000400,
  0x01000400, 0x00010400, 0x00010400, 0x01010000, 0x01010000, 0x01000404,
  0x00010004, 0x01000004, 0x01000004, 0x00010004, 0x00000000, 0x00000404,
  0x00010404, 0x01000000, 0x00010000, 0x01010404, 0x00000004, 0x01010000,
  0x01010400, 0x01000000, 0x01000000, 0x00000400, 0x01010004, 0x00010000,
  0x00010400, 0x01000004, 0x00000400, 0x00000004, 0x01000404, 0x00010404,
  0x01010404, 0x00010004, 0x01010000, 0x01000404, 0x01000004, 0x00000404,
  0x00010404, 0x01010400, 0x00000404, 0x01000400, 0x01000400, 0x00000000,
  0x00010004, 0x00010400, 0x00000000, 0x01010004
};

static const uint32_t sbox2[64] = {
  0x80108020, 0x80008000, 0x00008000, 0x00108020, 0x00100000, 0x00000020,
  0x80100020, 0x80008020, 0x80000020, 0x80108020, 0x80108000, 0x80000000,
  0x80008000, 0x00100000, 0x00000020, 0x80100020, 0x00108000, 0x00100020,
  0x80008020, 0x00000000, 0x80000000, 0x00008000, 0x00108020, 0x80100000,
  0x00100020, 0x80000020, 0x00000000, 0x00108000, 0x00008020, 0x80108000,
  0x80100000, 0x00008020, 0x00000000, 0x00108020, 0x80100020, 0x00100000,
  0x80008020, 0x80100000, 0x80108000, 0x00008000, 0x80100000, 0x80008000,
  0x00000020, 0x80108020, 0x00108020, 0x00000020, 0x00008000, 0x80000000,
  0x00008020, 0x80108000, 0x00100000, 0x80000020, 0x00100020, 0x80008020,
  0x80000020, 0x00100020, 0x00108000, 0x00000000, 0x80008000, 0x00008020,
  0x80000000, 0x80100020, 0x80108020, 0x00108000
};

static const uint32_t sbox3[64] = {
  0x00000208, 0x08020200, 0x00000000, 0x08020008, 0x08000200, 0x00000000,
  0x00020208, 0x08000200, 0x00020008, 0x08000008, 0x08000008, 0x00020000,
  0x08020208, 0x00020008, 0x08020000, 0x00000208, 0x08000000, 0x00000008,
  0x08020200, 0x00000200, 0x00020200, 0x08020000, 0x08020008, 0x00020208,
  0x08000208, 0x00020200, 0x00020000, 0x08000208, 0x00000008, 0x08020208,
  0x00000200, 0x08000000, 0x08020200, 0x08000000, 0x00020008, 0x00000208,
  0x00020000, 0x08020200, 0x08000200, 0x00000000, 0x00000200, 0x00020008,
  0x08020208, 0x08000200, 0x08000008, 0x00000200, 0x00000000, 0x08020008,
  0x08000208, 0x00020000, 0x08000000, 0x08020208, 0x00000008, 0x00020208,
  0x00020200, 0x08000008, 0x08020000, 0x08000208, 0x00000208, 0x08020000,
  0x00020208, 0x00000008, 0x08020008, 0x00020200
};

static const uint32_t sbox4[64] = {
  0x00802001, 0x00002081, 0x00002081, 0x00000080, 0x00802080, 0x00800081,
  0x00800001, 0x00002001, 0x00000000, 0x00802000, 0x00802000, 0x00802081,
  0x00000081, 0x00000000, 0x00800080, 0x00800001, 0x00000001, 0x00002000,
  0x00800000, 0x00802001, 0x00000080, 0x00800000, 0x00002001, 0x00002080,
  0x00800081, 0x00000001, 0x00002080, 0x00800080, 0x00002000, 0x00802080,
  0x00802081, 0x00000081, 0x00800080, 0x00800001, 0x00802000, 0x00802081,
  0x00000081, 0x00000000, 0x00000000, 0x00802000, 0x00002080, 0x00800080,
  0x00800081, 0x00000001, 0x00802001, 0x00002081, 0x00002081, 0x00000080,
  0x00802081, 0x00000081, 0x00000001, 0x00002000, 0x00800001, 0x00002001,
  0x00802080, 0x00800081, 0x00002001, 0x00002080, 0x00800000, 0x00802001,
  0x00000080, 0x00800000, 0x00002000, 0x00802080
};

static const uint32_t sbox5[64] = {
  0x00000100, 0x02080100, 0x02080000, 0x42000100, 0x00080000, 0x00000100,
  0x40000000, 0x02080000, 0x40080100, 0x00080000, 0x02000100, 0x40080100,
  0x42000100, 0x42080000, 0x00080100, 0x40000000, 0x02000000, 0x40080000,
  0x40080000, 0x00000000, 0x40000100, 0x42080100, 0x42080100, 0x02000100,
  0x42080000, 0x40000100, 0x00000000, 0x42000000, 0x02080100, 0x02000000,
  0x42000000, 0x00080100, 0x00080000, 0x42000100, 0x00000100, 0x02000000,
  0x40000000, 0x02080000, 0x42000100, 0x40080100, 0x02000100, 0x40000000,
  0x42080000, 0x02080100, 0x40080100, 0x00000100, 0x02000000, 0x42080000,
  0x42080100, 0x00080100, 0x42000000, 0x42080100, 0x02080000, 0x00000000,
  0x40080000, 0x42000000, 0x00080100, 0x02000100, 0x40000100, 0x00080000,
  0x00000000, 0x40080000, 0x02080100, 0x40000100
};

static const uint32_t sbox6[64] = {
  0x20000010, 0x20400000, 0x00004000, 0x20404010, 0x20400000, 0x00000010,
  0x20404010, 0x00400000, 0x20004000, 0x00404010, 0x00400000, 0x20000010,
  0x00400010, 0x20004000, 0x20000000, 0x00004010, 0x00000000, 0x00400010,
  0x20004010, 0x00004000, 0x00404000, 0x20004010, 0x00000010, 0x20400010,
  0x20400010, 0x00000000, 0x00404010, 0x20404000, 0x00004010, 0x00404000,
  0x20404000, 0x20000000, 0x20004000, 0x00000010, 0x20400010, 0x00404000,
  0x20404010, 0x00400000, 0x00004010, 0x20000010, 0x00400000, 0x20004000,
  0x20000000, 0x00004010, 0x20000010, 0x20404010, 0x00404000, 0x20400000,
  0x00404010, 0x20404000, 0x00000000, 0x20400010, 0x00000010, 0x00004000,
  0x20400000, 0x00404010, 0x00004000, 0x00400010, 0x20004010, 0x00000000,
  0x20404000, 0x20000000, 0x00400010, 0x20004010
};

static const uint32_t sbox7[64] = {
  0x00200000, 0x04200002, 0x04000802, 0x00000000, 0x00000800, 0x04000802,
  0x00200802, 0x04200800, 0x04200802, 0x00200000, 0x00000000, 0x04000002,
  0x00000002, 0x04000000, 0x04200002, 0x00000802, 0x04000800, 0x00200802,
  0x00200002, 0x04000800, 0x04000002, 0x04200000, 0x04200800, 0x00200002,
  0x04200000, 0x00000800, 0x00000802, 0x04200802, 0x00200800, 0x00000002,
  0x04000000, 0x00200800, 0x04000000, 0x00200800, 0x00200000, 0x04000802,
  0x04000802, 0x04200002, 0x04200002, 0x00000002, 0x00200002, 0x04000000,
  0x04000800, 0x00200000, 0x04200800, 0x00000802, 0x00200802, 0x04200800,
  0x00000802, 0x04000002, 0x04200802, 0x04200000, 0x00200800, 0x00000000,
  0x00000002, 0x04200802, 0x00000000, 0x00200802, 0x04200000, 0x00000800,
  0x04000002, 0x04000800, 0x00000800, 0x00200002
};

static const uint32_t sbox8[64] = {
  0x10001040, 0x00001000, 0x00040000, 0x10041040, 0x10000000, 0x10001040,
  0x00000040, 0x10000000, 0x00040040, 0x10040000, 0x10041040, 0x00041000,
  0x10041000, 0x00041040, 0x00001000, 0x00000040, 0x10040000, 0x10000040,
  0x10001000, 0x00001040, 0x00041000, 0x00040040, 0x10040040, 0x10041000,
  0x00001040, 0x00000000, 0x00000000, 0x10040040, 0x10000040, 0x10001000,
  0x00041040, 0x00040000, 0x00041040, 0x00040000, 0x10041000, 0x00001000,
  0x00000040, 0x10040040, 0x00001000, 0x00041040, 0x10001000, 0x00000040,
  0x10000040, 0x10040000, 0x10040040, 0x10000000, 0x00040000, 0x10001040,
  0x00000000, 0x10041040, 0x00040040, 0x10000040, 0x10040000, 0x10001000,
  0x10001040, 0x00000000, 0x10041040, 0x00041000, 0x00041000, 0x00001040,
  0x00001040, 0x00040040, 0x10000000, 0x10041000
};

/*
 * These two tables are part of the 'permuted choice 1' function.
 * In this implementation several speed improvements are done.
 */
static const uint32_t leftkey_swap[16] = {
  0x00000000, 0x00000001, 0x00000100, 0x00000101,
  0x00010000, 0x00010001, 0x00010100, 0x00010101,
  0x01000000, 0x01000001, 0x01000100, 0x01000101,
  0x01010000, 0x01010001, 0x01010100, 0x01010101
};

static const uint32_t rightkey_swap[16] = {
  0x00000000, 0x01000000, 0x00010000, 0x01010000,
  0x00000100, 0x01000100, 0x00010100, 0x01010100,
  0x00000001, 0x01000001, 0x00010001, 0x01010001,
  0x00000101, 0x01000101, 0x00010101, 0x01010101,
};

/*
 * Numbers of left shifts per round for encryption subkeys.  To
 * calculate the decryption subkeys we just reverse the ordering of
 * the calculated encryption subkeys, so there is no need for a
 * decryption rotate tab.
 */
static const unsigned char encrypt_rotate_tab[16] = {
  1, 1, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 1
};

/*
 * Macro to swap bits across two words.
 */
#define DO_PERMUTATION(a, temp, b, offset, mask)        \
    temp = ((a>>offset) ^ b) & mask;                    \
    b ^= temp;                                          \
    a ^= temp<<offset;


/*
 * This performs the 'initial permutation' of the data to be encrypted
 * or decrypted. Additionally the resulting two words are rotated one bit
 * to the left.
 */
#define INITIAL_PERMUTATION(left, temp, right)          \
    DO_PERMUTATION(left, temp, right, 4, 0x0f0f0f0f)    \
    DO_PERMUTATION(left, temp, right, 16, 0x0000ffff)   \
    DO_PERMUTATION(right, temp, left, 2, 0x33333333)    \
    DO_PERMUTATION(right, temp, left, 8, 0x00ff00ff)    \
    right =  (right << 1) | (right >> 31);              \
    temp  =  (left ^ right) & 0xaaaaaaaa;               \
    right ^= temp;                                      \
    left  ^= temp;                                      \
    left  =  (left << 1) | (left >> 31);

/*
 * The 'inverse initial permutation'.
 */
#define FINAL_PERMUTATION(left, temp, right)            \
    left  =  (left << 31) | (left >> 1);                \
    temp  =  (left ^ right) & 0xaaaaaaaa;               \
    left  ^= temp;                                      \
    right ^= temp;                                      \
    right  =  (right << 31) | (right >> 1);             \
    DO_PERMUTATION(right, temp, left, 8, 0x00ff00ff)    \
    DO_PERMUTATION(right, temp, left, 2, 0x33333333)    \
    DO_PERMUTATION(left, temp, right, 16, 0x0000ffff)   \
    DO_PERMUTATION(left, temp, right, 4, 0x0f0f0f0f)


/*
 * A full DES round including 'expansion function', 'sbox substitution'
 * and 'primitive function P' but without swapping the left and right word.
 * Please note: The data in 'from' and 'to' is already rotated one bit to
 * the left, done in the initial permutation.
 */
#define DES_ROUND(from, to, work, subkey)               \
    work = from ^ *subkey++;                            \
    to ^= sbox8[  work      & 0x3f ];                   \
    to ^= sbox6[ (work>>8)  & 0x3f ];                   \
    to ^= sbox4[ (work>>16) & 0x3f ];                   \
    to ^= sbox2[ (work>>24) & 0x3f ];                   \
    work = ((from << 28) | (from >> 4)) ^ *subkey++;    \
    to ^= sbox7[  work      & 0x3f ];                   \
    to ^= sbox5[ (work>>8)  & 0x3f ];                   \
    to ^= sbox3[ (work>>16) & 0x3f ];                   \
    to ^= sbox1[ (work>>24) & 0x3f ];

/*
 * Macros to convert 8 bytes from/to 32bit words.
 */
#define READ_64BIT_DATA(data, left, right)                                 \
    left  = (data[0] << 24) | (data[1] << 16) | (data[2] << 8) | data[3];  \
    right = (data[4] << 24) | (data[5] << 16) | (data[6] << 8) | data[7];

#define WRITE_64BIT_DATA(data, left, right)                                \
    data[0] = (left >> 24) &0xff; data[1] = (left >> 16) &0xff;    \
    data[2] = (left >> 8) &0xff; data[3] = left &0xff;                     \
    data[4] = (right >> 24) &0xff; data[5] = (right >> 16) &0xff;          \
    data[6] = (right >> 8) &0xff; data[7] = right &0xff;

/*
 * des_key_schedule():    Calculate 16 subkeys pairs (even/odd) for
 *                        16 encryption rounds.
 *                        To calculate subkeys for decryption the caller
 *                        have to reorder the generated subkeys.
 *
 *    rawkey:       8 Bytes of key data
 *    subkey:       Array of at least 32 uint32_ts. Will be filled
 *                  with calculated subkeys.
 *
 */
static void
des_key_schedule (const char * _rawkey, uint32_t * subkey)
{
  const unsigned char *rawkey = (const unsigned char *) _rawkey;
  uint32_t left, right, work;
  int round;

  READ_64BIT_DATA (rawkey, left, right)
    DO_PERMUTATION (right, work, left, 4, 0x0f0f0f0f)
    DO_PERMUTATION (right, work, left, 0, 0x10101010)
    left = ((leftkey_swap[(left >> 0) & 0xf] << 3)
            | (leftkey_swap[(left >> 8) & 0xf] << 2)
            | (leftkey_swap[(left >> 16) & 0xf] << 1)
            | (leftkey_swap[(left >> 24) & 0xf])
            | (leftkey_swap[(left >> 5) & 0xf] << 7)
            | (leftkey_swap[(left >> 13) & 0xf] << 6)
            | (leftkey_swap[(left >> 21) & 0xf] << 5)
            | (leftkey_swap[(left >> 29) & 0xf] << 4));

  left &= 0x0fffffff;

  right = ((rightkey_swap[(right >> 1) & 0xf] << 3)
           | (rightkey_swap[(right >> 9) & 0xf] << 2)
           | (rightkey_swap[(right >> 17) & 0xf] << 1)
           | (rightkey_swap[(right >> 25) & 0xf])
           | (rightkey_swap[(right >> 4) & 0xf] << 7)
           | (rightkey_swap[(right >> 12) & 0xf] << 6)
           | (rightkey_swap[(right >> 20) & 0xf] << 5)
           | (rightkey_swap[(right >> 28) & 0xf] << 4));

  right &= 0x0fffffff;

  for (round = 0; round < 16; ++round)
    {
      left = ((left << encrypt_rotate_tab[round])
              | (left >> (28 - encrypt_rotate_tab[round]))) & 0x0fffffff;
      right = ((right << encrypt_rotate_tab[round])
               | (right >> (28 - encrypt_rotate_tab[round]))) & 0x0fffffff;

      *subkey++ = (((left << 4) & 0x24000000)
                   | ((left << 28) & 0x10000000)
                   | ((left << 14) & 0x08000000)
                   | ((left << 18) & 0x02080000)
                   | ((left << 6) & 0x01000000)
                   | ((left << 9) & 0x00200000)
                   | ((left >> 1) & 0x00100000)
                   | ((left << 10) & 0x00040000)
                   | ((left << 2) & 0x00020000)
                   | ((left >> 10) & 0x00010000)
                   | ((right >> 13) & 0x00002000)
                   | ((right >> 4) & 0x00001000)
                   | ((right << 6) & 0x00000800)
                   | ((right >> 1) & 0x00000400)
                   | ((right >> 14) & 0x00000200)
                   | (right & 0x00000100)
                   | ((right >> 5) & 0x00000020)
                   | ((right >> 10) & 0x00000010)
                   | ((right >> 3) & 0x00000008)
                   | ((right >> 18) & 0x00000004)
                   | ((right >> 26) & 0x00000002)
                   | ((right >> 24) & 0x00000001));

      *subkey++ = (((left << 15) & 0x20000000)
                   | ((left << 17) & 0x10000000)
                   | ((left << 10) & 0x08000000)
                   | ((left << 22) & 0x04000000)
                   | ((left >> 2) & 0x02000000)
                   | ((left << 1) & 0x01000000)
                   | ((left << 16) & 0x00200000)
                   | ((left << 11) & 0x00100000)
                   | ((left << 3) & 0x00080000)
                   | ((left >> 6) & 0x00040000)
                   | ((left << 15) & 0x00020000)
                   | ((left >> 4) & 0x00010000)
                   | ((right >> 2) & 0x00002000)
                   | ((right << 8) & 0x00001000)
                   | ((right >> 14) & 0x00000808)
                   | ((right >> 9) & 0x00000400)
                   | ((right) & 0x00000200)
                   | ((right << 7) & 0x00000100)
                   | ((right >> 7) & 0x00000020)
                   | ((right >> 3) & 0x00000011)
                   | ((right << 2) & 0x00000004)
                   | ((right >> 21) & 0x00000002));
    }
}

void
gl_des_setkey (gl_des_ctx *ctx, const char * key)
{
  int i;

  des_key_schedule (key, ctx->encrypt_subkeys);

  for (i = 0; i < 32; i += 2)
    {
      ctx->decrypt_subkeys[i] = ctx->encrypt_subkeys[30 - i];
      ctx->decrypt_subkeys[i + 1] = ctx->encrypt_subkeys[31 - i];
    }
}

void
gl_des_ecb_crypt (gl_des_ctx *ctx, const char * _from, char * _to, int mode)
{
  const unsigned char *from = (const unsigned char *) _from;
  unsigned char *to = (unsigned char *) _to;
  uint32_t left, right, work;
  uint32_t *keys;

  keys = mode ? ctx->decrypt_subkeys : ctx->encrypt_subkeys;

  READ_64BIT_DATA (from, left, right)
    INITIAL_PERMUTATION (left, work, right)
    DES_ROUND (right, left, work, keys) DES_ROUND (left, right, work, keys)
    DES_ROUND (right, left, work, keys) DES_ROUND (left, right, work, keys)
    DES_ROUND (right, left, work, keys) DES_ROUND (left, right, work, keys)
    DES_ROUND (right, left, work, keys) DES_ROUND (left, right, work, keys)
    DES_ROUND (right, left, work, keys) DES_ROUND (left, right, work, keys)
    DES_ROUND (right, left, work, keys) DES_ROUND (left, right, work, keys)
    DES_ROUND (right, left, work, keys) DES_ROUND (left, right, work, keys)
    DES_ROUND (right, left, work, keys) DES_ROUND (left, right, work, keys)
    FINAL_PERMUTATION (right, work, left)
    WRITE_64BIT_DATA (to, right, left)
}
